File: examples/cache.zig
const std = @import("std");

/// A simple cache backed by a hash map.
pub fn Cache(comptime K: type, comptime V: type) type {
    return struct {
        const Self = @This();

        store: std.AutoHashMap(K, V),
        max_size: usize,

        pub fn init(allocator: std.mem.Allocator, max_size: usize) Self {
            return Self{
                .store = std.AutoHashMap(K, V).init(allocator),
                .max_size = max_size,
            };
        }

        pub fn deinit(self: *Self) void {
            self.store.deinit();
        }

        pub fn get(self: *const Self, key: K) ?V {
            return self.store.get(key);
        }

        pub fn put(self: *Self, key: K, value: V) !void {
            if (self.store.count() >= self.max_size) {
                var it = self.store.iterator();
                if (it.next()) |entry| {
                    _ = self.store.remove(entry.key_ptr.*);
                }
            }
            try self.store.put(key, value);
        }

        pub fn contains(self: *const Self, key: K) bool {
            return self.store.contains(key);
        }

        pub fn clear(self: *Self) void {
            self.store.clearRetainingCapacity();
        }
    };
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var cache = Cache([]const u8, i64).init(allocator, 100);
    defer cache.deinit();

    try cache.put("key1", 42);
    try cache.put("key2", 100);

    if (cache.get("key1")) |value| {
        std.debug.print("Found: {d}\n", .{value});
    }

    // TODO: Add TTL support
    // FIXME: Thread safety
}

File: examples/cache.rs
use std::collections::HashMap;

/// A simple cache implementation
pub struct Cache<K, V> {
    data: HashMap<K, V>,
    max_size: usize,
}

impl<K, V> Cache<K, V>
where
    K: Eq + std::hash::Hash,
{
    /// Create a new cache with the given maximum size
    pub fn new(max_size: usize) -> Self {
        Self {
            data: HashMap::new(),
            max_size,
        }
    }

    /// Insert a key-value pair into the cache
    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
        if self.data.len() >= self.max_size {
            self.data
                .keys()
                .next()
                .cloned()
                .and_then(|k| self.data.remove(&k));
        }
        self.data.insert(key, value)
    }

    /// Get a value from the cache
    pub fn get(&self, key: &K) -> Option<&V> {
        self.data.get(key)
    }

    /// Check if the cache contains a key
    pub fn contains(&self, key: &K) -> bool {
        self.data.contains_key(key)
    }

    /// Clear the cache
    pub fn clear(&mut self) {
        self.data.clear();
    }
}

fn main() {
    let mut cache = Cache::new(100);
    cache.insert("key1", 42);
    cache.insert("key2", 100);

    if let Some(value) = cache.get(&"key1") {
        println!("Found: {}", value);
    }

    // TODO: Add more sophisticated eviction policy
    // FIXME: Handle thread safety
    todo!()
}
